[NFC] Separate UnwindTable from DebugFrame into a different file

amsen20 opened the pull request:
By separating the Unwind table into a different file, this functionality can be a part of the DWARF library with no dependency on MC, which makes it usable in the MC layer.

This is a continuation of [PR#14520](https://github.com/llvm/llvm-project/pull/142520).

Commit:
Remove UnwindTable dependency on CIE, and FDE (transitively on llvm/Object)

For creating new UnwindTable, two static methods was implemented inside it, to create an instance of it from a CIE or FDE.
This static methods are moved out of the class as a library functions.

Commit:
Remove UnwindTable dependency on CIE, and FDE (transitively on llvm/Object)

For creating new UnwindTable, two static methods was implemented inside it, to create an instance of it from a CIE or FDE.
This static methods are moved out of the class as a library functions.

Commit:
Separate UnwindTable from DWARF Debug Frame

ilovepi reviewed (commented):
As in the other review, the C++ refactoring looks fine since its mostly mechanical, but this change is a decision code owners need to weigh in on. Fuller description of the change, rationale, and intended future work would go a long way to make the case for the refactor.

amsen20 commented:
> @Sterling-Augustine is this part of your DWARF expression parsing work? Or coincidentally overlapping with it?
> 
> @amsen20 where do you plan to put this code once it is refactored? Into MC itself, elsewhere, some new library?

AFAIK @Sterling-Augustine 's refactoring is towards a new low-level DWARF library that does not depend on Object and MC. That library would be usable from MC and Object or any other library in LLVM. This will also be located in the same library that for example `CFIProgram` will be.

Commit:
Annotate the extracted UnwindTable creators with LLVM ABI

Commit:
Annotate UnwindTable exported API with LLVM_ABI

Commit:
Annotate exported API of DebugFrame with LLVM_ABI

Commit:
Merge branch 'main' into unwind-table-separation

Commit:
Redo the separation with the updates

dwblaikie commented on llvm/include/llvm/DebugInfo/DWARF/DWARFUnwindTable.h:
```diff
@@ -0,0 +1,378 @@
+/// A class that represents a location for the Call Frame Address (CFA) or a
+/// register. This is decoded from the DWARF Call Frame Information
+/// instructions and put into an UnwindRow.
+class UnwindLocation {
```
Should this patch include the deletion of the original/previous version of this, so this patch shows up more as a move and less as a copy/paste?

amsen20 replied:
I guess your comment is on the wrong commit, after the [final commit](https://github.com/llvm/llvm-project/pull/142521/commits/fbadafcb02f1a25fa96a4e5e0b3644ddc241e336), they are deleted from `DWARFDebugFrame.h`. I guess the file changes in the PR is also correct.

dwblaikie replied:
Might've just had this open at an old version of the patch or something - or github was jus tbeing weird. I went back and forth to check a few times, and it seemed to only show 7 commits and looked as though these pieces were duplicated, not moved.

but it seems to be showing something better now...

Commit:
Merge branch 'main' into unwind-table-separation

Commit:
Use the new expression printer function

Commit:
Lint

amsen20 commented:
@dwblaikie
@igorkudrin

The whole point of this NFC was to get `UnwindTable` into the low-level DWARF library. Now that the [NFC](https://github.com/llvm/llvm-project/pull/145081) creating this library has been merged, separating it in the same library doesn't make sense and has to be moved into the new library.

I tried moving `UnwindTable` to the new library but encountered one problem `UnwindTable` depends on `DWARFExpressionPrinter`.
This newly created file is separated from `DWARFExpression` which takes care of printing stuff.

If you agree I have to do the same for the `UnwindTable` printing stuff:
- Create a new file called `UnwindTablePrinter`
- Move the `UnwindLocation::dump` method to `UnwindTablePrinter` has a static function that gets an `UnwindLocation` instance as an argument and does the printing stuff.

I can do this plan in this PR, or I can separate the files here, and open another PR that does the moving.

Please let me know.

amsen20 commented on llvm/include/llvm/DebugInfo/DWARF/DWARFUnwindTable.h:
```diff
@@ -0,0 +1,374 @@
+//===- DWARFUnwindTable.h ----------------------------------------*- C++-*-===//
```
I tried to not change any actual code of the current existing files. If this is the case, is this change exempted from that?

amsen20 replied:
I changed the new header file comments to match the new style.

igorkudrin commented:
> @dwblaikie @igorkudrin
> 
> The whole point of this NFC was to get `UnwindTable` into the low-level DWARF library. Now that the [NFC](https://github.com/llvm/llvm-project/pull/145081) creating this library has been merged, separating it in the same library doesn't make sense and has to be moved into the new library.
> 
> I tried moving `UnwindTable` to the new library but encountered one problem `UnwindTable` depends on `DWARFExpressionPrinter`. This newly created file is separated from `DWARFExpression` which takes care of printing stuff.
> 
> If you agree I have to do the same for the `UnwindTable` printing stuff:
> 
> * Create a new file called `UnwindTablePrinter`
> * Move the `UnwindLocation::dump` method to `UnwindTablePrinter` has a static function that gets an `UnwindLocation` instance as an argument and does the printing stuff.
> 
> I can do this plan in this PR, or I can separate the files here, and open another PR that does the moving.
> 
> Please let me know.

I honestly believe that in good design, printing and dumping functionality should be separated from the class itself. Like, nobody expects `std::vector` to dump its contents, but rather provide enough methods to do so externally in any necessary way. So I'm totally in favor of the splitting. I think this PR is a good place for this work.

Commit:
Merge branch 'main' into unwind-table-separation

Commit:
Merge main

Commit:
Lint

Commit:
[TEMP] move unwind table to where it should be, the layering is not correct

Commit:
Move the printing logic of UnwindTable to UnwindTablePrinter

Commit:
Lint

amsen20 commented:
@dwblaikie
@igorkudrin

I moved all the printing logic to `UnwindTablePrinter`. Through the way, I have added a couple of accessor methods.
Please check the code again.

To make the reviewing process, I first moved the `UnwindTable` to the low-level library in this [commit](https://github.com/llvm/llvm-project/pull/142521/commits/1ed0df26166199c0081f54b06f4d606f34c01834) and then separated the printing logic.

Thanks.

amsen20 committed:
Merge branch 'main' into unwind-table-separation

amsen20 commented on llvm/include/llvm/DebugInfo/DWARF/LowLevel/DWARFUnwindTable.h:
```diff
@@ -0,0 +1,321 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_DEBUGINFO_DWARF_DWARFUNWINDTABLE_H
+#define LLVM_DEBUGINFO_DWARF_DWARFUNWINDTABLE_H
```
It is not followed in all headers in that folder, should I edit them too?

amsen20 replied:
Fixed for this PR and added a new [PR#147172](https://github.com/llvm/llvm-project/pull/147172)

amsen20 committed:
Remove unnecessary print functions from library

amsen20 commented on llvm/include/llvm/DebugInfo/DWARF/DWARFUnwindTablePrinter.h:
```diff
@@ -0,0 +1,102 @@
+/// conversion because some register numbers differ between the two sections
+/// for certain architectures like x86.
+LLVM_ABI void printUnwindLocation(const UnwindLocation &R, raw_ostream &OS,
+                                  DIDumpOptions DumpOpts);
```
Fixed

amsen20 committed:
Fix header guards match folder paths

amsen20 commented on llvm/include/llvm/DebugInfo/DWARF/LowLevel/DWARFUnwindTable.h:
```diff
@@ -0,0 +1,321 @@
+
+} // end namespace llvm
+
+#endif
```
Fixed.

igorkudrin reviewed (approved):
LGTM. Please update the headers for the new files as @ilovepi suggested. Please wait a couple of days in case @dwblaikie has anything to add.

igorkudrin commented:
> All the `DebugInfo/DWARF` do not follow this new header style, should I apply this style to this new files or all of them?

You are only responsible for the code and files that you are adding or changing. To keep your commit atomic and focused, you should not update anything unrelated to the purpose of the PR. However, if you wish, you can prepare a separate NFC PR to align other files with the current Coding Standards. Of course, you are not obliged to do so.

amsen20 committed:
Sync new file header with new style

igorkudrin commented:
cpp files also should be updated

amsen20 committed:
Match starting docs for cpp files with new style guide

amsen20 commented:
> cpp files also should be updated

Sorry missed them, fixed
